ledger/participant-integration-api/src/main/scala/platform/store/backend/postgresql/PostgresStorageBackend.scala

```diff
@@ -11,706 +11,53 @@ import com.daml.platform.store.backend.{DBDTOV1, StorageBackend}
-        batch.create_agreement_text,
-        batch.create_key_value,
-        batch.create_key_hash,
-        batch.template_id,
-        batch.flat_event_witnesses,
-        batch.tree_event_witnesses,
-        batch.event_sequential_id,
-        batch.create_argument_compression,
-        batch.create_key_value_compression,
-      )
-    )
-
-    rawDBBatchPostgreSQLV1.eventsBatchConsumingExercise.foreach(batch =>
-      execute(
-        preparedInsertEventsBatchConsumingExercise,
-        batch.event_id,
-        batch.event_offset,
-        batch.contract_id,
-        batch.transaction_id,
-        batch.ledger_effective_time,
-        batch.node_index,
-        batch.command_id,
-        batch.workflow_id,
-        batch.application_id,
-        batch.submitters,
-        batch.create_key_value,
-        batch.exercise_choice,
-        batch.exercise_argument,
-        batch.exercise_result,
-        batch.exercise_actors,
-        batch.exercise_child_event_ids,
-        batch.template_id,
-        batch.flat_event_witnesses,
-        batch.tree_event_witnesses,
-        batch.event_sequential_id,
-        batch.create_key_value_compression,
-        batch.exercise_argument_compression,
-        batch.exercise_result_compression,
-      )
-    )
-
-    rawDBBatchPostgreSQLV1.eventsBatchNonConsumingExercise.foreach(batch =>
-      execute(
-        preparedInsertEventsBatchNonConsumingExercise,
-        batch.event_id,
-        batch.event_offset,
-        batch.contract_id,
-        batch.transaction_id,
-        batch.ledger_effective_time,
-        batch.node_index,
-        batch.command_id,
-        batch.workflow_id,
-        batch.application_id,
-        batch.submitters,
-        batch.create_key_value,
-        batch.exercise_choice,
-        batch.exercise_argument,
-        batch.exercise_result,
-        batch.exercise_actors,
-        batch.exercise_child_event_ids,
-        batch.template_id,
-        batch.flat_event_witnesses,
-        batch.tree_event_witnesses,
-        batch.event_sequential_id,
-        batch.create_key_value_compression,
-        batch.exercise_argument_compression,
-        batch.exercise_result_compression,
-      )
-    )
-
-    rawDBBatchPostgreSQLV1.packageEntriesBatch.foreach(batch =>
-      execute(
-        preparedInsertPackageEntryBatch,
-        batch.ledger_offset,
-        batch.recorded_at,
-        batch.submission_id,
-        batch.typ,
-        batch.rejection_reason,
-      )
-    )
-
-    rawDBBatchPostgreSQLV1.packagesBatch.foreach(batch =>
-      execute(
-        preparedInsertPackageBatch,
-        batch.package_id,
-        batch.upload_id,
-        batch.source_description,
-        batch.size,
-        batch.known_since,
-        batch.ledger_offset,
-        batch._package,
-      )
-    )
+    def executeTable(pgTable: PGTable[_], data: Array[Array[_]]): Unit =
+      if (data(0).length > 0) execute(pgTable.insertStatement, pgTable.setupData(data, _))
```

Comment:
Why isn't `data(0)` throwing?

Reply:
Or is `data(0)` some sort of header that's always present?

Reply:
the meaning of this two dimensional array: 
* 1st dimension (outer array): table columns
* 2nd dimension: to-be-inserted data for those table columns (this will be unnest-ed)

so data(0) means: the array of data to be inserted for the first column
note: if data(0) would be out of bonds, that would mean we have a table setup without columns
note2: all 2nd dimension arrays are of the same length (transposed structure)

:thinking: actually I will add a comment to explain why it is safe, good point